### PR TITLE
Add ProducerStateTable::count()/clear()

### PR DESCRIPTION
**Pull request by jipanyang**
…has been picked up by consumer.

Signed-off-by: Jipan Yang <jipan.yang@alibaba-inc.com>

### CODE REVIEW

**Comment by jipanyang** [comment]
@nikos-github suggested  another solution to clear the keyset and temporary table for producerState, it should work too.   I may create another PR for that if needed.

**Comment by rodnymolina** [comment]
@jipanyang thanks for taking care of this one. It looks good to me.

**Comment by zhenggen-xu** [comment]
@jipanyang  As talked, it is probably good to have another function to erase the data in the keyset and temporary table for the particular producerState, so the caller won't need wait for it to be empty.  In case of doing reconciliation, as long as the data is consistent in the consumer and database (plain table), later entries from applications would be passed through and got to the right state.

File: common/producerstatetable.cpp [hunk 1]
```diff
@@ -122,4 +122,15 @@ void ProducerStateTable::flush()
...
 }
 
+bool ProducerStateTable::isEmpty()
+{
+    RedisCommand cmd;
+    cmd.format("SCARD %s", getKeySetName().c_str());
+    RedisReply r = m_pipe->push(cmd);
+    r.checkReplyType(REDIS_REPLY_INTEGER);
+
+    long long int setSize = r.getReply<long long int>();
```
Comment (qiluo-msft) [inline common/producerstatetable.cpp]: >setSize [](start = 18, length = 7)

Maybe it's more useful to rename the function as count, and return the setSize directly.

**Comment by rodnymolina** [comment]
@zhenggen-xu, good idea. However, in that case i don't see a need to call isEmpty() first -- one single call to (e.g) deletePending() would be enough.

File: common/producerstatetable.cpp [hunk 2]
```diff
@@ -122,15 +127,39 @@ void ProducerStateTable::flush()
     m_pipe->flush();
 }
 
-bool ProducerStateTable::isEmpty()
+int64_t ProducerStateTable::setSize()
```
Comment (rodnymolina) [inline common/producerstatetable.cpp]: i think you meant to say: getSize()

Comment (rodnymolina) [reply common/producerstatetable.cpp]: Never mind, I see your point. Disregard.

File: common/producerstatetable.cpp [hunk 3]
```diff
@@ -122,4 +127,39 @@ void ProducerStateTable::flush()
...
+    cmd.format("SCARD %s", getKeySetName().c_str());
+    RedisReply r = m_pipe->push(cmd);
+    r.checkReplyType(REDIS_REPLY_INTEGER);
+
+    return r.getReply<long long int>();
+}
+
+// Warning: calling this function will cause all data in keyset and the temporary table to be abandoned.
+// ConsumerState may have got the notification from PUBLISH, but will see no data popped.
+void ProducerStateTable::drop()
```
Comment (qiluo-msft) [inline common/producerstatetable.cpp]: >drop [](start = 25, length = 4)

How about data pushed in the PUB/SUB channel between ProducerStateQueue and ConsumerStateQueue?

File: common/producerstatetable.cpp [hunk 4]
```diff
@@ -122,4 +127,39 @@ void ProducerStateTable::flush()
...
+{
+    // Assembly redis command args into a string vector
+    vector<string> args;
+    args.emplace_back("EVALSHA");
+    args.emplace_back(m_shaDrop);
+    args.emplace_back("2");
+    args.emplace_back(getKeySetName());
+    args.emplace_back(getStateHashPrefix() + getTableName());
+    args.emplace_back("G");
+    args.emplace_back("''");
```
Comment (qiluo-msft) [inline common/producerstatetable.cpp]: I guess they are copied. necessary here?

File: common/producerstatetable.h [hunk 5]
```diff
@@ -26,12 +26,16 @@ class ProducerStateTable : public TableBase, public TableName_KeySet
 
     void flush();
 
+    int64_t pendingCount();
+
+    void drop();
```
Comment (qiluo-msft) [inline common/producerstatetable.h]: By design, ProducerStateTable and ProducerTable have the exact same interfaces. Do you mind also implementing for the other?

File: common/producerstatetable.cpp [hunk 6]
```diff
@@ -122,4 +127,39 @@ void ProducerStateTable::flush()
     m_pipe->flush();
 }
 
+int64_t ProducerStateTable::pendingCount()
```
Comment (qiluo-msft) [inline common/producerstatetable.cpp]: >pendingCount [](start = 28, length = 12)

I think 'count' is a better name. For a queue concept, there is nothing pending.

File: common/producerstatetable.cpp (hunk 3 above)
Comment (qiluo-msft) [inline common/producerstatetable.cpp]: >drop [](start = 25, length = 4)

Call it clear()? like deque::clear()

Comment (jipanyang) [reply common/producerstatetable.cpp]: If you know how to safely revoke the data already  "PUBLISH", please advise.

File: common/producerstatetable.cpp (hunk 4 above)
Comment (jipanyang) [reply common/producerstatetable.cpp]: place holder for argv.  If people don't mind script like
```
       "redis.call('DEL', KEYS[1])\n"
        "redis.call('DEL', ARGV[1])\n";
```
I could remove them.  But hope no back and forth.

File: common/producerstatetable.h (hunk 5 above)
Comment (jipanyang) [reply common/producerstatetable.h]: This is not a "solve all your problems" PR,  it should be better do that in separate PR.

File: common/producerstatetable.cpp (hunk 6 above)
Comment (jipanyang) [reply common/producerstatetable.cpp]: Ok.  Again, hope no back and forth.

File: common/producerstatetable.cpp (hunk 3 above)
Comment (jipanyang) [reply common/producerstatetable.cpp]: Ok.  Again, hope no back and forth.

File: common/producerstatetable.cpp (hunk 4 above)
Comment (qiluo-msft) [reply common/producerstatetable.cpp]: I mean 2 lines actually.
```
    args.emplace_back("G");
    args.emplace_back("''");
```

---
In reply to: [214504794](https://github.com/Azure/sonic-swss-common/pull/227#discussion_r214504794) [](ancestors = 214504794)

Comment (jipanyang) [reply common/producerstatetable.cpp]: Yes, both lines are placeholder for argv.  Two keys are specified, putting two argv, no real use for now.    I thinks it is a trivial case, please just indicate what you want,   two place holder and rename "G" to ''  / one place holder/no argv place holder/one key one argv?

Comment (qiluo-msft) [reply common/producerstatetable.cpp]: I prefer no placeholders at all. If not possible, one ore two empty string as placeholders are ok.

**Pull request merged**